cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/AvroFileManager.java

```diff
@@ -0,0 +1,248 @@
+    return logFileOutputStream;
+  }
+
+  private LogFileOutputStream createAvroFile(LogPathIdentifier identifier, long timestamp) throws IOException {
+    Location location = getLocation(identifier, timestamp);
+    LOG.info("Creating Avro file {}", location);
+    LogFileOutputStream logFileOutputStream = new LogFileOutputStream(location);
+    try {
+      fileMetaDataManager.writeMetaData(identifier, timestamp, location);
+    } catch (Throwable e) {
+      closeAndDelete(logFileOutputStream);
+      throw new IOException(e);
+    }
+
+    fileMap.put(identifier, logFileOutputStream);
+    return logFileOutputStream;
+  }
+
+  private LogFileOutputStream rotateFile(LogFileOutputStream logFileOutputStream,
+                                         LogPathIdentifier identifier, long timestamp) throws IOException {
+    long currentTs = System.currentTimeMillis();
+    long timeSinceFileCreate = currentTs - logFileOutputStream.getCreateTime();
+    if (timeSinceFileCreate > maxLifetimeMillis) {
+      logFileOutputStream.close();
+      return createAvroFile(identifier, timestamp);
+    }
+    return logFileOutputStream;
+  }
+
+  /**
+   * closes all the open avro files in the map
+   */
+  public void close() {
+    // close all the files in fileMap
+    // clear the map
+    for (LogFileOutputStream file : fileMap.values()) {
+      try {
+        file.flush();
+      } catch (Exception e) {
+        LOG.debug("Exception while flushing contents to file {}", file.getLocation(), e);
+      } finally {
+        Closeables.closeQuietly(file);
+      }
+    }
+    fileMap.clear();
+  }
+
+  private void closeAndDelete(LogFileOutputStream logFileOutputStream) {
+    try {
+      try {
+        logFileOutputStream.close();
+      } finally {
+        if (logFileOutputStream.getLocation().exists()) {
+          logFileOutputStream.getLocation().delete();
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Error while closing and deleting file {}", logFileOutputStream.getLocation(), e);
+    }
+  }
+
+  /**
+   * flushes the contents of all the open avro files
+   * @throws IOException
+   */
+  public void flush() throws IOException {
+    // perform flush on all the files in the fileMap
+    for (LogFileOutputStream file : fileMap.values()) {
+      try {
+        file.flush();
+      } catch (Exception e) {
+        LOG.debug("Exception while flushing contents to file {}", file.getLocation());
+        Closeables.closeQuietly(file);
+        throw Throwables.propagate(e);
+      }
+    }
+  }
+
+  /**
+   * Represents an Avro file.
+   *
+   * Since there is no way to check the state of the underlying file on an exception,
+   * all methods of this class assume that the file state is bad on any exception and close the file.
+   */
+  class LogFileOutputStream implements Closeable {
+    private final Location location;
+    private FSDataOutputStream outputStream;
+    private DataFileWriter<GenericRecord> dataFileWriter;
```

Comment:
if i keep this final, then i am not able to close on exception in constructor, as it complains that it was might not have initialized.